Fix bugcrowd adhoc stability issues and improve default configuration

unlox775-code-dot-org opened the pull request:
# Fix bugcrowd adhoc stability issues and improve default configuration

<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->

This PR addresses three stability and configuration issues for adhoc environments:

1. **[CANCELLED]** ~~**Bugcrowd instance size upgrade (INF-1945)**: The bugcrowd adhoc environment was experiencing periodic hangs and 499 errors due to CPU saturation on the `t3.2xlarge` burstable instance. Upgraded to `m7i.4xlarge` for sustained CPU performance.~~
   - Instead, we will use the `INSTANCE_TYPE=` command line argument.

3. **Debug pages fix (INF-1770)**: Fixed the logic that disables debug pages on bugcrowd adhoc. The check was using exact equality (`!= 'bugcrowd'`) but the stack name is `'adhoc-bugcrowd'`, so debug pages were incorrectly enabled.

4. **Rolling restart default**: Changed the default for `enableWebServiceProcessRollingRestart` from `false` to `true`. This workaround for memory leaks has been enabled manually in production for years, and all environments that run for extended periods need this. Making it the default ensures new environments automatically get this protection without manual configuration.

## Links
<!--  Links to relevant external resources.
      - spec docs, Jira tickets, related PRs, Honeybadger errors, etc.
-->

- Jira: [INF-1945](https://codedotorg.atlassian.net/browse/INF-1945) - Bugcrowd adhoc environment instability
- Jira: [INF-1770](https://codedotorg.atlassian.net/browse/INF-1770) - Debug pages showing on bugcrowd adhoc

## Testing story
<!--  Does your change include appropriate tests?
      - If so, please describe how the tests included in this PR are sufficient.
      - If not, please explain why this change does not need to be tested. -->

- **Instance size change**: Will be tested when the bugcrowd stack is recreated (done). CloudFormation template validation passes.
- **Debug pages fix**: Logic change is straightforward - checks if stack name includes 'bugcrowd' instead of exact equality. Can be verified after deployment by triggering an error and confirming debug pages don't show.
   - Cannot test this until after this deploys to production, in which case this new code change will also update in production. And then we may have to restart the dashboard database. But at that point, debug pages should be disabled on the ad hoc environment, closing issue INF-1770.
- **Rolling restart default**: This is a configuration default change. Existing environments with the flag manually set will continue to use their configured value. New environments will default to enabled, matching production behavior.

## Follow-up work
<!--  List any clean-up or technical debt that will be addressed in future work.
      - Include Jira links where applicable.
      - Are there any known limitations or improvements planned? -->

- Monitor bugcrowd adhoc environment after instance upgrade to confirm CPU utilization improves and 499 errors decrease.
- This change of rolling restart = `true` is essentially accepting our memory leak as not going to be fixed. It unfortunately is at a very low priority for us to fix right now and not really in sight for the sake of technical debt prioritization. This unfortunately means that we are accepting the memory leak, but nevertheless, this eventually should be fixed, at which point this default could be set back to `false`.

## Security
<!--  How does this change impact our security surface-area?
      - Do API's touched have the right auth?
      - Do infra changes impact our attack surface or encryption?
-->

- **Debug pages fix**: Security improvement - prevents sensitive error information from being exposed on bugcrowd adhoc environment.

## PR Creation Checklist:
<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [X] Security impacts have been documented
- [X] Code is well-commented
- [ ] Relevant documentation has been added or updated
- [X] User impact is well-understood and desirable
- [X] Follow-up work items (including potential tech debt) are tracked and linked

unlox775-code-dot-org committed:
Fix bugcrowd adhoc stability issues and improve default configuration

- Upgrade bugcrowd adhoc instance from t3.2xlarge to m7i.4xlarge (INF-1945)
- Fix debug pages showing on bugcrowd adhoc by using substring check (INF-1770)
- Enable rolling restarts by default for all environments

Addresses INF-1945 and INF-1770

cat5inthecradle commented on aws/cloudformation/cloud_formation_stack.yml.erb:
```diff
@@ -59,7 +59,7 @@ Parameters:
   InstanceType:
     Type: String
-    Default: <%=rack_env?(:production) ? 'm7i.12xlarge' : 't3.2xlarge'%>
+    Default: <%=rack_env?(:production) ? 'm7i.12xlarge' : (stack_name.include?('bugcrowd') ? 'm7i.4xlarge' : 't3.2xlarge')%>
```
Thinking ahead to the next person that wants a big adhoc, there are probably cleaner ways to write this.

`stack_name.match?(/bugcrowd|makemebigtoo/)` maybe? Or we parameterize some kind of "XL Adhoc" feature? I definitely want to avoid it getting *more* complicated than you've currently got it.

sureshc replied:
This is a Stack Parameter. For one-off exceptions like Bugcrowd, we can just change the value of the Parameter on this one Stack instead of changing the code that sets the default value. There's even a way to set the Parameter values at the time we execute the rake `adhoc:start` command from an Environment variable.

unlox775-code-dot-org replied:
I think related to Elijah's commit below that I think it is a good idea to actually mention the bug crowd ad hoc completely by name (potentially even in a `rack_env?` or something) because like it or not the bugcrowd-adhoc is one of our permanent fixture environments.  That puts it in a different category as just other (temporary) ad hoc environments that people want to make that are huge.

So I think finding a way to do this the right way is not a bad idea. The nested ternary operator is not pretty. Do you think I should abstract it to an actual function somewhere? I would say yes if I wasn't trying to lean more away from ERBs that are tied to external Ruby code, at least not until we have more clear standards of the right way to do this.

sureshc replied:
I think I'd prefer to document that we provision adhocs with:
 `bundle exec rake adhoc:cdn:start RAILS_ENV=adhoc STACK_NAME=adhoc-bugcrowd DATABASE=true INSTANCE_TYPE= m7i.4xlarge`

Instead of having code in the template that sets the default for the bugcrowd environment. If we change vendors for security research, do we want to be updating hard coded references in this template? This ERB logic just sets the Default, and we are free to specify a different value than the default.

unlox775-code-dot-org replied:
Agreed, I will remove this and use `INSTANCE_TYPE=` instead, thanks!

Hamms commented on dashboard/config/environments/adhoc.rb:
```diff
@@ -15,7 +15,7 @@
   # Consider disabling full error reports for profiling/load-testing by setting
   # this flag to false, due to memory leak: https://github.com/rails/rails/issues/27273
-  config.consider_all_requests_local = ENV['STACK_NAME'] != 'bugcrowd'
+  config.consider_all_requests_local = ENV['STACK_NAME'].to_s.exclude?('bugcrowd')
```
I don't think we want to apply this logic to _all_ adhocs with `bugcrowd` in the name; for example, an adhoc created off of this branch.

Any reason not to do:

```suggestion
  config.consider_all_requests_local = ENV['STACK_NAME'] != 'adhoc-bugcrowd'
```

(both here and in `cloud_formation_stack.yml.erb` above)

unlox775-code-dot-org replied:
Agreed. I don't know why I was just doing a partial string match anyways.  It's updated now.

Hamms commented on lib/cdo/app_server_hooks.rb:
```diff
@@ -22,7 +22,9 @@ def self.before_fork
-      if Gatekeeper.allows('enableWebServiceProcessRollingRestart')
+      # Default to enabled (true) since this is needed in production and all environments to work around memory leaks.
+      # Can still be disabled via Gatekeeper if needed.
+      if Gatekeeper.allows('enableWebServiceProcessRollingRestart', default: true)
```
This feels like a potentially impactful change; particularly given that most of our environments don't currently suffer negative effects from not having this gem enabled, and that its use is [strongly discouraged in all situations](https://github.com/zombocom/puma_worker_killer?tab=readme-ov-file#-stop-)

Are we sure we don't want to apply a more targeted change here in order to get this enabled on the bugcrowd server?

unlox775-code-dot-org replied:
I think my main thought and just turning it on by default is as long as it is on in production by default for right now, it's better for other things to be more production like in the meantime. I think initially it was thought we would just have it on for a short turn during an hour of code, but then it's just been on for a long time since.

I initially actually implemented it as a change to just effect bugcrowd adhoc, but then I decided this pivot maybe made more sense. What do you think?

unlox775-code-dot-org committed:
Use exact equality check for adhoc-bugcrowd instead of substring match

Addresses PR feedback - only apply special logic to the specific
adhoc-bugcrowd stack, not any adhoc with 'bugcrowd' in the name.

unlox775-code-dot-org committed:
Update cloud_formation_stack.yml.erb
